state/spacesdiscovery.go

```diff
@@ -71,13 +76,81 @@ func (st *State) SaveSubnetsFromProvider(subnets []network.SubnetInfo) error {
 			ProviderId:        subnet.ProviderId,
 			ProviderNetworkId: subnet.ProviderNetworkId,
 			CIDR:              subnet.CIDR,
+			SpaceName:         spaceName,
 			VLANTag:           subnet.VLANTag,
 			AvailabilityZone:  firstZone,
 		})
 		if err != nil {
 			return errors.Trace(err)
 		}
+
+	}
+
+	// We process FAN subnets separately for clarity.
+	cfg, err := st.ModelConfig()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	fans, err := cfg.FanConfig()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(fans) == 0 {
+		return nil
+	}
+
+	for _, subnet := range subnets {
+		_, subnetNet, err := net.ParseCIDR(subnet.CIDR)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, fan := range fans {
+			subnetSize, _ := subnetNet.Mask.Size()
+			underlaySize, _ := fan.Underlay.Mask.Size()
+			// We need to cut a part of fan specific for this physical subnet, eg:
+			// for FAN 172.31/16 -> 243/8 and physical subnet 172.31.64/20
+			// we get FAN subnet 243.64/12.
+			if underlaySize <= subnetSize && fan.Underlay.Contains(subnetNet.IP) {
+				id := fmt.Sprintf("%s-INFAN-%d-%d-%d-%d-%d", subnet.ProviderId, subnetNet.IP[0], subnetNet.IP[1], subnetNet.IP[2], subnetNet.IP[3], subnetSize)
+				if modelSubnetIds.Contains(id) {
+					continue
+				}
+				overlaySize, _ := fan.Overlay.Mask.Size()
+				newOverlaySize := overlaySize + (subnetSize - underlaySize)
+				fanSize := uint(underlaySize - overlaySize)
+				newFanIP := subnetNet.IP.To4()
+				for i := 0; i < 4; i++ {
+					newFanIP[i] &^= fan.Underlay.Mask[i]
+				}
+				numIp := binary.BigEndian.Uint32(newFanIP)
+				numIp <<= fanSize
+				binary.BigEndian.PutUint32(newFanIP, numIp)
+				for i := 0; i < 4; i++ {
+					newFanIP[i] += fan.Overlay.IP[i]
+				}
+				newOverlay := net.IPNet{IP: newFanIP, Mask: net.CIDRMask(newOverlaySize, 32)}
+				var firstZone string
+				if len(subnet.AvailabilityZones) > 0 {
+					firstZone = subnet.AvailabilityZones[0]
+				}
```

Comment:
splitting out all of this so that there is a simple helper function that can determine what the carved out address ranges are based on the grouping, and we have good tests for that function with edge cases, etc.
I'd put that func in 'network' probably.